BUGFIX: Make grow formulas faster and more accurate.

d0sboots opened the pull request:
This also adds a test for accuracy of the formulas. It is notable that the test does *not* pass on the previous version of the code. In particular, `numCyclesForGrowthCorrected` failed occasionally but catastrophically once values got above 1.44e20.

This also gets rid of a lot of annoying duplicate code, which is good because these functions can be relied upon to return identical answers now.

d0sboots committed:
Make grow formulas faster and more accurate.

This also adds a test for accuracy of the formulas. It is notable that
the test does *not* pass on the previous version of the code. In
particular, `numCyclesForGrowthCorrected` failed occaisonally but
catastrophically once values got above 1.44e20.

steven-r commented on src/Constants.ts:
```diff
@@ -125,8 +125,8 @@ export const CONSTANTS: {
   HomeComputerMaxRam: 1073741824, // 2 ^ 30
-  ServerBaseGrowthRate: 1.03, // Unadjusted Growth rate
-  ServerMaxGrowthRate: 1.0035, // Maximum possible growth rate (max rate accounting for server security)
+  ServerBaseGrowthIncr: 0.03, // Unadjusted growth increment (growth rate is 1 + this * adjustment)
```
I would suggest writing `(1 + this * adjustment)`

d0sboots replied:
I'm not quite sure what you mean... would the whole thing would become "(growth rate is (1 + this * adjustment))", with two parens?

steven-r replied:
I just looked at the comment --> 1 + this * adjustment is different to (1 + this) * adjustment

d0sboots replied:
Now I'm even more confused... Are you saying you read it with the wrong order of operations initially?

Edit: Does this make it more clear?

d0sboots committed:
Small doc improvements

d0sboots committed:
Simplify initial guess

This slightly speeds up the calculation and greatly simplifies the
explanation.

d0sboots committed:
Use a more accurate guess to slightly improve performance.

Add tests for a set of more challenging cases.
Allow using the fast-path directly from iteration 1 in more cases.

d0sboots committed:
Tweak the starting guess slightly, again.

This is definitely getting to maximum optimization now, unless a
radically new approach is found.

Also add more info to a task that was flaky, so we can hopefully
diagnose it in the future.
